### PR TITLE
adds loggerwrapper and emoji files

### PR DESCRIPTION
**Pull request by tiinanguyen**
This PR has the following changes:
* Creates the `LogIntent` file in Swift to add emojis to log messages. 
* Creates `Emoji` file.
* Updates `RCLogUtils.h` to use methods in `Logger`.
* Extends `Logger` for additional log methods, hidden behind a feature flag.
* Removes `RCReceiptInUseByOtherSubscriberError`, which is not mapped to a backend error.
* Switch case for backend errors to add emojis.

Files in `Purchases`:  `RCUserLog(@"%@", RCStrings.configure.initial_app_user_id)`
Files in `PurchasesCoreSwift`: `LoggerWrapper.user(Strings.configure.initial_app_user_id)`

### Requirements: 
- [x] ~3.9.2 released #423~

### CODE REVIEW

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (tiinanguyen) [inline PurchasesCoreSwift/Logging/LoggerWrapper.swift:44]: Following Android implementation, this is for logs that not necessarily an error, but the source is from Apple. Example: "could not find SKProduct for..." and "App running in sandbox without a receipt file. Restoring transactions won't work..."

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (tiinanguyen) [inline PurchasesCoreSwift/Logging/LoggerWrapper.swift:48]: Also following Android implementation. Examples: "setCreative called" and "Debug logging enabled"

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (aboedo) [inline PurchasesCoreSwift/Logging/LoggerWrapper.swift:21]: I wouldn't name this `emojis`, since the important part of it isn't really the emoji - it's the prefix. 
i.e.: we could functionally replace all `‼️` with `IMPORTANT` tomorrow and the code would still make sense, because the emoji is an implementation detail. 
so I'd name this `suffix` so that it also conveys how it's meant to be used. 

for bonus swiftiness, it could also be 
```swift
var suffix: String { 
```
(the rest of the code would remain the same, just a signature change). That way usage doesn't need `()`

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (aboedo) [inline PurchasesCoreSwift/Logging/LoggerWrapper.swift:11]: do we intend to use this directly from objective-c? if not, the class shouldn't be public (internal is the default), and the `@objc` can be ommitted

File: PurchasesCoreSwift/Logging/Strings/Emojis.swift
Comment (aboedo) [inline PurchasesCoreSwift/Logging/Strings/Emojis.swift:11]: `Emojis` -> `Emoji` (singular) would be standard

File: PurchasesCoreSwift/Logging/Strings/Emojis.swift
Comment (aboedo) [inline PurchasesCoreSwift/Logging/Strings/Emojis.swift:12]: these aren't really emojis, we're attaching meaning to them. 
I think the _meaning_ or intent should be a responsibility of `LogIntent`, since that's what it does, and this enum should just be a mapping of emoji to unicode, i.e.: 

```
enum Emoji: String { 
    case apple = "\u{1F34E}\u{FE0F}"
    case heartCatEyes = "\u{1F63B}\u{FE0F}"
    ...
```

Note that this also removes the need for comments on each line, since it's perfectly clear what each one does

Comment (aboedo) [reply PurchasesCoreSwift/Logging/Strings/Emojis.swift:12]: and then `LogIntent` would do something like: 

```swift
switch self {
         case .appleError: return Emoji.apple.rawValue + Emoji.doubleExclamation.rawValue
```

Which makes total sense, it's essentially saying `appleError = apple + !!`

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (aboedo) [inline PurchasesCoreSwift/Logging/LoggerWrapper.swift:38]: I'd call this `fullMessage` or `messageWithPrefix`

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (aboedo) [inline PurchasesCoreSwift/Logging/LoggerWrapper.swift:36]: I think we could just make this a method of Logger instead of having a separate wrapper, since there aren't clear gains from having two separate public classes

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (aboedo) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:11]: we might also wanna have a separate file for it, we mostly try to keep 1 file per class

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (aboedo) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:48]: I think we might wanna just replace the `debug` method in `Logger` with this one, and we could keep the name `debug` so that there are no other code changes needed

Comment (aboedo) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:48]: that way all existing calls to `RCDebugLog` automatically gain the prefix

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (aboedo) [inline PurchasesCoreSwift/Logging/LoggerWrapper.swift:65]: should this be info? it feels pretty eventful for just a debug level

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (aboedo) [inline PurchasesCoreSwift/Logging/LoggerWrapper.swift:77]: similar comment as with `debugInfo`, we might just want to use this instead of the current version of `warn`

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (aboedo) [inline PurchasesCoreSwift/Logging/LoggerWrapper.swift:53]: similar thought here

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (aboedo) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:44]: maybe those should be of warning level?

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (aboedo) [inline PurchasesCoreSwift/Logging/LoggerWrapper.swift:61]: what are examples of these?

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (tiinanguyen) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:44]: since we want developers to know the source of the error, would this just be another LogIntent case with `appleWarn = apple + warning` or cluster these logs with `warning`?

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (tiinanguyen) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:48]: I was planning on merging this PR to work on the other string files, but then other developers would see the info emoji before releasing debug logs. Should this be refactored after all string files are merged?

File: PurchasesCoreSwift/Logging/Strings/Emojis.swift
Comment (tiinanguyen) [reply PurchasesCoreSwift/Logging/Strings/Emojis.swift:12]: This makes sense, I'll also make updates to the Android version to keep implementation consistent

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (tiinanguyen) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:61]: Some examples are "Subscriber attributes errors...", "The App User ID is not valid.", and "Error restoring purchase"

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (tiinanguyen) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:77]: same comment as above. I was planning on merging this PR to work on the other string files, but then other developers would see the emojis before releasing debug logs. Should this be refactored after all string files are merged?

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (tiinanguyen) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:48]: On second thought, I could create another branch, something like "[FINAL] debug logs" and merge all string files into that branch. Once we're ready for release merge that into develop. Thoughts?

Comment (tiinanguyen) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:48]: On second thought, I could create another branch called "[FINAL] debug logs" and merge all changes (this PR + strings file) into there. Once we're ready for release we would merge that branch into develop. Thoughts?

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (tiinanguyen) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:36]: Similar to Android, I was thinking this file would be the wrapper for RCLogger. Instead of a separate log method, `appleError` could be implemented like this: 

```
@objc public static func appleError(_ message: String) {
        let messageWithPrefix = "\(LogIntent.appleError.suffix) \(message)"
        Logger.log(level: .error, message: messageWithPrefix)
    }
```
Would this implementation make more sense to justify the wrapper file?

Comment (aboedo) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:36]: My thinking is that these methods might replace a lot of the stuff in Logger, which also means devs don't have to figure out which of the two to use for a given situation. 

re: mix and match of objc and non-objc stuff, I think that's fine, it's basically just a distinction between public and internal stuff. 
I would maybe separate them with pragmas or better yet extensions so that they're organized

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (aboedo) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:44]: I meant doing something like `log.appleWarning("skus can't be found")`, so that we have a way for them to stand out in the logs

File: PurchasesCoreSwift/Logging/LoggerWrapper.swift
Comment (aboedo) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:48]: if we're worried about sending out a half-implemented version of this to the world, perhaps we could use a feature flag in order to have the whole code easy to test for us but not actually used by customers. 
as in, 

```swift
@objc public static func log(level: LogLevel, intent: LogIntent, message: String) {
#if V2_LOGS_ENABLED
    let emojifiedMessage = "\(intent.emojis()) \(message)"
    Logger.log(level: level, message: emojifiedMessage)
#else 
    Logger.log(level: level, message: message)
#end
}
```

Comment (tiinanguyen) [reply PurchasesCoreSwift/Logging/LoggerWrapper.swift:48]: I tried replacing the `debug` method in `Logger`, but `RCDebugLog` calls aren't going into the `log(level: LogLevel, intent: LogIntent, message: String)` method. I placed a breakpoint at `RCDebugLog(@"Debug logging enabled.");` in RCPurchases and clicking into the function brought me to the `log(level: LogLevel, message: String)` function. 

Also, where do I define the custom compiler flag? I tried adding it to the target's build settings under Preprocessor Macros and tried Swift Compiler - Custom Flags, but the `V2_LOGS_ENABLED` don't populate when I try a test call of `[RCLogger appleError: @"test"]`

File: PurchasesCoreSwift/Logging/Logger.swift
Comment (tiinanguyen) [inline PurchasesCoreSwift/Logging/Logger.swift:33]: @aboedo Error from CircleCI: "cannot find type 'LogIntent' in scope" but I'm able to compile and run this code on a simulator. This also happened when this function was an `@objc public static func`.

Comment (aboedo) [reply PurchasesCoreSwift/Logging/Logger.swift:33]: `LogIntent.swift` is in the file system and in the repo, but it hasn't been added to the `.xcproject` file, so xcode can't find it. Dragging and dropping it from Finder to the project in Xcode should fix it, same for `Emoji.swift`. 
Make sure they get added to the target, i.e.:
image

Comment (aboedo) [reply PurchasesCoreSwift/Logging/Logger.swift:33]: maybe the changes to the `.xcproject` files have been done but not committed? it'd explain why it works for you locally

File: PurchasesCoreSwift/Logging/Strings/Emoji.swift
Comment (aboedo) [inline PurchasesCoreSwift/Logging/Strings/Emoji.swift:19]: I think the comments aren't needed here anymore, the emoji names seem descriptive enough

**Review by aboedo — COMMENTED** [review]
looking great! just gotta fix the `xcodeproj` thing

File: PurchasesCoreSwift/Logging/Logger.swift
Comment (aboedo) [inline PurchasesCoreSwift/Logging/Logger.swift:59]: this probably needs to be `@objc public` in order for obj-c to see it

Comment (tiinanguyen) [reply PurchasesCoreSwift/Logging/Logger.swift:59]: Adding the `@objc public` decorator could be applied to the extension or to the methods in the extension, either or would work. I will go with the former.

**Review by aboedo — APPROVED** [review]
looks great! let's hold off on merging just for a bit since the 3.9.2 release is being prepared and this is for 3.10.0. I'll add a HOLD tag which we can remove as soon as 3.9.2 is out

**Comment by aboedo** [comment]
@tiinanguyen HOLD removed, this can be merged whenever you're ready

File: PurchasesCoreSwift/Logging/Strings/Emoji.swift
Comment (aboedo) [inline PurchasesCoreSwift/Logging/Strings/Emoji.swift:12]: this should probably have occurred to me sooner, but can't you just use the emoji instead of the unicode? @tiinanguyen

Comment (tiinanguyen) [reply PurchasesCoreSwift/Logging/Strings/Emoji.swift:12]: @aboedo Ah, yeah emojis can be used here instead of the unicode. I did a test run just to be sure. Can I just push a commit into `develop` without opening a PR?

Comment (aboedo) [reply PurchasesCoreSwift/Logging/Strings/Emoji.swift:12]: a PR is always better, even if only for a sanity check. I know it's a hassle, but it should just take a min, and it wouldn't block any of the other PRs

**Pull request merged**
